src/addrman.cpp

```diff
@@ -815,7 +815,7 @@ std::vector<CAddress> AddrManImpl::GetAddr_(size_t max_addresses, size_t max_pct
 
     size_t nNodes = vRandom.size();
     if (max_pct != 0) {
-        nNodes = max_pct * nNodes / 100;
+        nNodes = std::min(nNodes, max_pct * nNodes / 100);
```

Comment:
Since there was a bug somewhere if we ever send `max_pct > 100` (with privacy implications, the reason `max_pct` exists is not to reveal the entire addrman), why not `Assume(max_pct <=100)` / adjust the fuzzer range instead of trying to accommodate it? There is a difference between `max_pct` and the `max_addresses` where a caller could legitimately ask for more addresses than addrman has because they don't know current addrman size - no caller should ever ask for more than 100%.

Reply:
I am fine either way and slightly prefer the softer approach of capping it to 100 instead of asserting that it is <=100. There was a discussion some time ago about crashing the program way too easily and reserving the assert/crash only for cases where the continuation of the program really does not make sense. I kind of agree with that.

Yes, `Assume()` is some middle ground between `assert()` and the softer approach. `Assume()` + cap in release builds looks reasonable to me too. Plus document it properly: "make sure you never pass >100 or you will crash the program" ... hmm, could this end up in all callers doing the cap at the call site before calling `GetAddr()` to make sure to avoid a crash?

Do you remember the bug number? "The reason max_pct exists is not to reveal the entire addrman", but then passing even == 100 is a bug?

---

If you will be touching this, there is this minor mostly theoretical thing: `max_pct * nNodes` could overflow and I guess `max_pct * nNodes / 100` could end up with a lower value than `nNodes` for some very high input `max_pct`. So, if you will be touching this, maybe it is "safer" to implement the cap as:

```cpp
max_pct = std::min(max_pct, 100);
nNodes = max_pct * nNodes / 100;
```

(this occurred to me just recently)

Reply:
> Yes, Assume() is some middle ground between assert() and the softer approach. Assume() + cap in release builds looks reasonable to me too.

I agree. I can address it. 

> so in current master nothing but 23 or 0 should ever be passed.

So, we could adjust both addrman and connman fuzz targets to get 23 or 0 to use for `max_pct`. Sounds good?

Reply:
> I think fuzzing in range between 0 and 100 would make sense - after all that's what the function is designed to handle

I agree.

Reply:
Done.